content/blog/2019-09-25-vril-vector-bril.md

```diff
@@ -0,0 +1,139 @@
+- evaluating performance of `vadd`
+
+#### Integration tests for Vectors
+We added tests to generate JSON and text representation of Bril from each other to gain confidence about its translatability. These tests are included in `vtest/parse` and `vtest/print`.
+
+#### Functional testing of `vadd`
+We wrote a program with `vadd` and ran this test in `vrili` with different values for `maxvl` which represents the size of vector lenght of the backend. The test generated expected results. This is included in `vtest/interp/array_vector.bril`.
+
+#### Performance of `vadd`
+Initial motivation for Vril is to reduce hop counts by having vector operators in Bril. This aspect is tested by comparing dynamic instruction count, hop count and lines of code for a loop executed in a scalar manner and a vector operation. 
+
+| Program           | Lines of code | Dynamic instructions | Hop count |
+|-------------------|---------------|----------------------|-----------|
+| Conventional Bril | 43            | 41                   | 1         |
+| Scalar Array      | 50            | 109                  | 12        |
+| Vector Array      | 48            | 44                   | 3         |
+
+Given that the length of array is N=10 and vector length selected by Vrili at execution is k, and compensating for the start and the end block as shown in the figure, scalar operation iterates N times and vector operation iterates N/k times.
+
+CFG for scalar array operation  | CFG for vector array operation
+:-------------------------:|:-------------------------:
+![](https://github.com/sa2257/vril/blob/82e9da0eed50a98fc5285a32020b0ba59385e398/plots/array_scalar_nv.png) | ![](https://github.com/sa2257/vril/blob/82e9da0eed50a98fc5285a32020b0ba59385e398/plots/array_vector_nv.png)
+
+Conclusion
+--------------------------------------
+We have extended bril to support array types. We have added two new operations to move data in and out of the arrays so that we can emulate data movements between an array and what it would be a scalar register. We have also extended bril to support vector operations of two types: configuration and arithmetic. Configuration operations allow to modify a vector state and arithmetic operations perform operations on array arguments.
+
+The goal of this exercise was to understand how much an IR needs to change in order to express data-parallel operations. For that, we extended bril to be able to express vector operations and to compare their potential against a traditional scalar set of operations. For that we written a benchmark for vector-vector add (vvadd), which adds the elements of two arrays and stores their results into a third array in two versions: a scalar code and a vector code. The end goal is to verify that the CFG generated is very similar for both codes, it should contain the same number of basic blocks (BB). However, the vector code hops on the BB involved in the `for-loop` statement *vector length* times less than the scalar code. 
```

Comment:
This paragraph suggests that I still don't quite understand the role of the vector length—I thought that `vadd` added together two entire arrays? But this paragraph seems to suggests it adds together only an N-sized chunk of the array, where N is the vector length.

Reply:
From the programmer's point of view it will add two entire arrays. However, if the architecture does not provide sufficient vector length to do so, it will do this in chunks of N, where N is the architecture's vector length, i.e. `maxvl` in Vrili.

Reply:
The fact that it happens in chunks of N is in programmer's view. This N is captured at runtime in `vl` to iterate in the loop. Actual N is microarchitectural and is specified in `maxvl`.